demos/ota/aws_iot_ota_update_demo.c

```diff
@@ -334,92 +358,141 @@ static void App_OTACompleteCallback( OTA_JobEvent_t eEvent )
 {
-    int xRet = EXIT_SUCCESS;
-    IotMqttError_t xMqttInitStatus;
-    bool bMqttInitialized = false;
+    OTA_State_t eState;
+    static OTA_ConnectionContext_t xOTAConnectionCtx;
+
+    configPRINTF( ( "OTA demo version %u.%u.%u\r\n",
+                    xAppFirmwareVersion.u.x.ucMajor,
+                    xAppFirmwareVersion.u.x.ucMinor,
+                    xAppFirmwareVersion.u.x.usBuild ) );
 
-    /* Unused parameters */
-    ( void ) awsIotMqttMode;
-    ( void ) pIdentifier;
-    ( void ) pNetworkServerInfo;
+    retryInterval = OTA_DEMO_CONN_RETRY_BASE_INTERVAL_SECONDS;
 
-    if( otaDemoNETWORK_TYPES == AWSIOT_NETWORK_TYPE_NONE )
+    for( ; ; )
     {
-        configPRINTF( ( "There are no networks configured for the demo.\r\n" ) );
-        xRet = EXIT_FAILURE;
-    }
+        configPRINTF( ( "Connecting to broker...\r\n" ) );
+
+        /* Establish a new MQTT connection. */
+        if( _establishMqttConnection( awsIotMqttMode,
+                                      pIdentifier,
+                                      pNetworkServerInfo,
+                                      pNetworkCredentialInfo,
+                                      pNetworkInterface,
+                                      &mqttConnection ) == EXIT_SUCCESS )
+        {
+            /* Update the connection context shared with OTA Agent.*/
+            xOTAConnectionCtx.pxNetworkInterface = ( void * ) pNetworkInterface;
+            xOTAConnectionCtx.pvNetworkCredentials = pNetworkCredentialInfo;
+            xOTAConnectionCtx.pvControlClient = mqttConnection;
 
-    /* Create semaphore to notify network available */
-    if( xRet == EXIT_SUCCESS )
-    {
-        xNetworkAvailableLock = xSemaphoreCreateBinary();
+            /* Update the connection available flag.*/
+            networkConnected = pdTRUE;
 
-        if( xNetworkAvailableLock == NULL )
-        {
-            configPRINTF( ( "Failed to create semaphore.\r\n" ) );
-            xRet = EXIT_FAILURE;
-        }
-    }
+            /* Check if OTA Agent is suspended and resume.*/
+            if( ( eState = OTA_GetAgentState() ) == eOTA_AgentState_Suspended )
+            {
+                OTA_Resume( &xOTAConnectionCtx );
+            }
 
-    if( xRet == EXIT_SUCCESS )
-    {
-        /* Initialize the MQTT library. */
-        xMqttInitStatus = IotMqtt_Init();
+            /* Initialize the OTA Agent , if it is resuming the OTA statistics will be cleared for new connection.*/
+            OTA_AgentInit( ( void * ) ( &xOTAConnectionCtx ),
+                           ( const uint8_t * ) ( clientcredentialIOT_THING_NAME ),
+                           App_OTACompleteCallback,
+                           ( TickType_t ) ~0 );
 
-        if( xMqttInitStatus != IOT_MQTT_SUCCESS )
-        {
-            xRet = EXIT_FAILURE;
+            while( ( ( eState = OTA_GetAgentState() ) != eOTA_AgentState_Stopped ) && networkConnected )
```

Comment:
Where is `networkConnected` set to `pdFALSE`?  I notice it's not declared `volatile`.  If some other task is setting it, some compilers might optimize this test away.

Reply:
_networkConnected is set false in disconnect callback,making it volatile.